lib/api/traversing.js

```diff
@@ -12,6 +12,46 @@ var uniqueSort = require('htmlparser2').DomUtils.uniqueSort;
 var isTag = utils.isTag;
 var reSiblingSelector = /^\s*[~+]/;
 
+function _matcher(fn, reverse) {
+  // customized Map, discards null elements
+  function matchMap(elems, cb, arg) {
+    var len = elems.length;
+    var value;
+    var i = 0;
+    var ret = [];
+    for (; i < len; i++) {
+      value = cb(elems[i], i, arg);
+      if (value !== null) {
+        ret.push(value);
+      }
+    }
+    return Array.prototype.concat.apply([], ret);
+  }
+
+  return function (selector) {
+    if (this[0]) {
+      var matched = matchMap(this, fn, selector);
+
+      // select.filter uses uniqueSort already internally
+      if (typeof selector === 'string') {
+        matched = select.filter(selector, matched, this.options);
+      } else {
+        matched = uniqueSort(
```

Comment:
None of the previous methods seem to be using `uniqueSort`, which will lead to a slowdown here. Maybe uniqueness could be an option to this function, and done using a linear search as before?

Reply:
Yes, `prev*` and `next*` functions wont need sorting, but since `select.filter` uses same sorting internally it messes up order, if filter is used.

Reply:
also function `children` is unique but for all other functions uniqueSort should only used if original cheerio object length would be more than 1.

Since if you take `parents` from one tag each element would be unique and reverse sorted, but from two and up elements start to differ and one point start to be duplicate, also order may be harder to control. Same with most other functions.